src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CartBundle\EventListener;
+
+use Sylius\Component\Cart\Context\CartContextInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+final class SessionCartSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var CartContextInterface
+     */
+    private $cartContext;
+
+    /**
+     * @var string
+     */
+    private $sessionKeyName;
+
+    /**
+     * @param CartContextInterface $cartContext
+     * @param string $sessionKeyName
+     */
+    public function __construct(CartContextInterface $cartContext, $sessionKeyName)
+    {
+        $this->cartContext = $cartContext;
+        $this->sessionKeyName = $sessionKeyName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            KernelEvents::RESPONSE => ['onKernelResponse'],
+        ];
+    }
+
+    /**
+     * @param FilterResponseEvent $event
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        /** @var Request $request */
+        $request = $event->getRequest();
+        // Hacky hack. Until there is a better solution.
+        if (!$this->isHtmlRequest($request)) {
```

Comment:
> Nope nope nope nope nope...

Do you mean that part? 🎉 

Excluding / including anything by route sounds really, really bad and should be avoided at all costs (just ask @lchrusciel about his romance with SiteAccess :)). 

We already have channel contexts reorganised and they're working like a charm - the channel context always throws an exception when in AdminBundle (you can see that by looking at debug toolbar while browsing the admin panel).

We can use that to our advantage and make cart providers return a new cart (`NewCartProvider` in this PR) only if there is a channel found. So `CartProvider` / `CartContext` will always throw an exception if trying to get a current cart.

If there's no cart passed here, we cannot persist it in session :)